search/floydwarshall.go

```diff
@@ -156,38 +157,62 @@ func (p ShortestPaths) Between(u, v graph.Node) (path []graph.Node, weight float
 	}
 	path = []graph.Node{p.nodes[from]}
 	unique = true
+	seen := make([]int, len(p.nodes))
+	for i := range seen {
+		seen[i] = -1
+	}
+	seen[from] = 0
 	for from != to {
 		c := p.at(from, to)
 		if len(c) != 1 {
 			unique = false
 		}
-		from = c[rand.Intn(len(c))]
+		i := rand.Intn(len(c))
+		from = c[i]
+		if seen[from] >= 0 {
```

Comment:
Is there anything that prevents (repeatedly) take the same zero-weight cycle? There may theoretically be many such cycles to take from a node but only one non-cycle path to advance the algorithm.

Here if such a cycle is detected, `path` is reverted, but on the next iteration `from` is advanced randomly to the next node on a shortest path which may lie on the same cycle. The path will be trimmed again but `from` will be already on the cycle. Is my understanding correct or am I missing something?